plutus-tx/README.md

```diff
@@ -1,3 +1,228 @@
+        localWorks = plusOneLocal 1
+        -- You can of course bind this to a name, but for the purposes
+        -- of this tutorial we won't since TH requires it to be in
+        -- another module.
+        thWorks = $$([|| \(x::Int) -> x + 1 ||]) 1
+    in localWorks + thWorks
+    ||])
+```
+
+From this point on we're going to start dealing with more advanced features of
+Haskell, like datatypes. The way these are encoded into Plutus Core is quite
+tricky, so we're going to stop looking at the generated Plutus Core code, but
+it's in there if you're curious.
+
+TODO: when we store the PIR, start showing PIR at this stage (or earlier).
+
+We can use normal Haskell datatypes and pattern matching freely:
+
+```haskell
+matchMaybe :: PlcCode
+matchMaybe = $$(plutus [|| \(x:: Maybe Int) -> case x of
+    Just n -> n
+    Nothing -> 0
+   ||])
+```
+
+This works for your own datatypes too! (See [Haskell language support](#haskell-language-support)
+for some caveats.) Unlike functions, datatypes do not need to be defined inside the
+expression, hence why we can use types like `Maybe` from the `Prelude`.
+
+### Using PlutusTx builtins
+
+PlutusTx has some builtin types and functions available, both for working with primitive
+data (integers and bytestrings), and also for performing chain-specific operations.
+
+The PlutusTx builtins are available via the `Language.PlutusTx.Builtins` module. You
+shouldn't need to use the integer and bytestring builtins directly, these are mapped
+to the corresponding Haskell operations on `Int` and lazy `ByteString` directly. However,
+you may wish to use some of the others.
+
+`error` deserves a special mention. `error` causes the transaction to abort when it is
+evaluated, which is the way that validation failure is signalled.
+
+### Lifting values
+
+This is all very good for defining pieces of code *statically*, but you are
+likely to want to *dynamically* produce Plutus Core programs. For example, you
+might be writing the body of a transaction to initiate a crowdfunding smart contract,
+which would need to be parameterized by user input determining the size of the goal,
+the campaign start and end times, etc.
+
+You can do this by writing the static code as a function, and then passing an
+argument at runtime by *lifting* it and then applying the two programs together. As a
+very simple example, let's write an add-one function.
+
+```haskell
+-- TODO: show PIR here too
+
+-- |
+-- >>> let program = addOneToN 4
+-- >>> prettyPlcDef program
+-- (program 1.0.0
+--   [
+--     [
+--       (lam
+--         addInteger
+--         (fun [(con integer) (con 8)] (fun [(con integer) (con 8)] [(con integer) (con 8)]))
+--         (lam ds [(con integer) (con 8)] [ [ addInteger ds ] (con 8 ! 1) ])
+--       )
+--       { (builtin addInteger) (con 8) }
+--     ]
+--     (con 8 ! 4)
+--   ]
+-- )
+
+-- >>> prettyPlcDef $ runCk program
+-- (con 8 ! 5)
+addOneToN :: Int -> Program TyName Name ()
```

Comment:
This is the first time the `Program TyName Name ()` type shows up - it should be explained, OR there should be a version of `applyProgram` that just operates on the serialised code, similar to `Wallet.UTXO.Types.applyScript`

Reply:
Maybe the `Script` type, and `lifted` and `applyScript` should live in `plutus-tx`

Reply:
Yeah, I was also wondering about that. That's one of the infelicities at the moment. The thing is, I don't really want to have `applyProgram` on `PlcCode`, because that's supposed to be a serialized blob. And `Script` is pretty much just a wrapper around `Program` that's opaque for the sake of `wallet-api`. I'm not sure.

Reply:
TBH this whole bit looks terrible and I need to think about the interface. I'm going to leave it as it is here, but I'm definitely aware it needs improvement.